torch/_isl.py

```diff
@@ -0,0 +1,179 @@
+
+    # These functions are a straight port from tvm.cc in c2isl, and not very
+    # Pythonic.
+    #
+    # TODO: Delete these from the builder once it's easier to stitch operators
+    # together.
+    def makeMatMult(self, I, W, outputName, transI=False, transW=False):
+        outputShape = (I.shape[1 if transI else 0], W.shape[0 if transW else 1])
+        reductionExpr = I.shape[0 if transI else 1]
+        return self.tvmOpMakeMatMult(outputShape, [reductionExpr], I, W, outputName, transI, transW)
+
+    def access(self, input, i, j, transpose=False):
+        if transpose:
+            return input[j, i]
+        else:
+            return input[i, j]
+
+    def tvmOpMakeMatMult(self, outputShape, reductionExpr, input, weight, name, transposeInput, transposeWeight):
+        assert len(reductionExpr) == 1
+        r = reductionExpr[0]
+        k = tvm.reduce_axis((0, r), "k")
+
+        def l(i, j):
+            return tvm.sum(self.access(input, i, k, transposeInput) *
+                           self.access(weight, k, j, transposeWeight),
+                           axis=k)
+        # Register the op!
+        return self.compute(outputShape, l, name=name)
+
+
+class IslFunction(object):
+    """
+    Superclass for operators which are defined directly using TVM
+    IR and then compiled into runnable CUDA code using c2isl.
+    See IslMatMul for an example of how to instantiate this class.
+
+    Limitations:
+      * CUDA only
+      * As TVM IR requires you to specify the types of placeholder
+        variables upfront, this means that operators created
+        with this are necessarily monomorphic.
+      * If you reuse this operator, you must call it again with
+        tensors that are exactly the same size (this is because
+        c2isl generates size-specialized kernels).
+    """
+
+    # This is an irritating little function.  I used the pre-existing
+    # plumbing for exposing C++ Functions as Python Functions.
+    # But these are not "true" classes: they're just functions that
+    # return a Python wrapper to the underlying C++ object. This
+    # means we can't subclass them object.  So instead we overload
+    # __new__ to do all of the necessary setup, and then dispatch over.
+    # Ouch!  Perhaps worth a rewrite if we decide to do the C++ end
+    # properly.
+    def __new__(cls, *args, **kwargs):
+        builder = TvmBuilder()
+        cls.tvm(builder, *args, **kwargs)
+
+        def swizzle(ls):
+            return tuple(l.handle.value for l in ls)
+        return torch._C._functions.IslFunction(
+            cls.name(),
+            *map(swizzle, (builder.outputs, builder.inputs, builder.vars, builder.ops)))
+
+    @staticmethod
+    def tvm(builder):
+        raise NotImplementedError
+
+    @staticmethod
+    def name():
+        raise NotImplementedError
+
+
+# This is an example use-case of this functionality.
+class IslMatMul(IslFunction):
+    @staticmethod
+    def name():
+        return "matmul"
```

Comment:
Is there any reason why we want to make it a method instead of a class attribute (`name = "matmul"`), or just use the class name?